File: matrixMultiplicationC++/matrixMultiplication.cpp
#include "matrixMultiplication.hpp"



void readArr(string fileName, int *arr)
{
    std::ifstream file(fileName);

    int count = 0;

    if (file.is_open()) {
        std::string line;
        while (getline(file, line)) {
            std::stringstream ss(line);
            int num;
            while (ss >> num) {
                arr[count] = num;
                count++;
                if (ss.peek() == ',') {
                    ss.ignore();
                }
            }
        }
        file.close();
    } else {
        std::cout << "Unable to open file.\n";
    }
}

void matrixMultiplication(int *firstMatrix, int *secondMatrix, long *resultMatrix, int matrixSize)
{
    for (int i = 0; i < matrixSize; i++) {
        for (int j = 0; j < matrixSize; j++) {
            long sum = 0;
            for (int k = 0; k < matrixSize; k++) {
                sum += firstMatrix[i * matrixSize + k] * secondMatrix[k * matrixSize + j];
            }
            resultMatrix[i * matrixSize + j] = sum;
        }
    }
}


void writeResult(long *resultArray, int size){
    std::ofstream outFile("result.txt");

    if (outFile.is_open()) {
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                outFile << resultArray[i * size + j];
                if (j < size - 1) {
                    outFile << ",";
                }
            }
            outFile << "\n";
        }
        outFile.close();
    } else {
        std::cout << "Unable to open file.\n";
    }
}

File: matrix_multiplication_rustUnsafe/src/matrix_multiplication.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Result, Write};

pub fn read_arr(file_name: &str, arr: &mut [i64], matrix_size: usize) -> Result<()> {
    let file = File::open(file_name)?;
    let reader = BufReader::new(file);

    for (i, line) in reader.lines().enumerate() {
        for (j, num) in line?.split(',').enumerate() {
            let index = i * matrix_size + j;
            arr[index] = num.trim().parse().unwrap();
        }
    }
    Ok(())
}

pub unsafe fn matrix_multiplication(
    first_matrix: &[i64],
    second_matrix: &[i64],
    result_matrix: &mut [i64],
    matrix_size: usize,
) {
    for i in 0..matrix_size {
        for j in 0..matrix_size {
            let mut sum = 0;
            for k in 0..matrix_size {
                let a = *first_matrix.get_unchecked(i * matrix_size + k);
                let b = *second_matrix.get_unchecked(k * matrix_size + j);
                sum += a * b;
            }
            *result_matrix.get_unchecked_mut(i * matrix_size + j) = sum;
        }
    }
}

pub fn write_result(file_name: &str, result_array: &[i64], size: usize) -> Result<()> {
    let mut file = File::create(file_name)?;
    for i in 0..size {
        let mut row_string = String::new();
        for j in 0..size {
            row_string.push_str(&result_array[i * size + j].to_string());
            if j < size - 1 {
                row_string.push(',');
            }
        }
        file.write_all(row_string.as_bytes())?;
        file.write_all(b"\n")?;
    }
    Ok(())
}